File: dsa-problems/Queues/solutions/cpp/QUE-007-lab-window-instability.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <set>

using namespace std;

class Solution {
public:
    vector<long long> windowInstability(const vector<int>& values, int k) {
        int n = values.size();
        vector<long long> result;
        deque<int> minD, maxD;
        
        // Use multiset for median (simplest in C++ as it supports deletion)
        // Two multisets: small (max), large (min)
        multiset<int> small, large;
        
        auto balance = [&]() {
            while (small.size() > large.size() + 1) {
                large.insert(*small.rbegin());
                small.erase(prev(small.end()));
            }
            while (large.size() > small.size()) {
                small.insert(*large.begin());
                large.erase(large.begin());
            }
        };
        
        auto add = [&](int val) {
            small.insert(val);
            large.insert(*small.rbegin());
            small.erase(prev(small.end()));
            balance();
        };
        
        auto remove = [&](int val) {
            auto it = small.find(val);
            if (it != small.end()) {
                small.erase(it);
            } else {
                large.erase(large.find(val));
            }
            balance();
        };
        
        for (int i = 0; i < n; i++) {
            // Min/Max Deques
            while (!minD.empty() && minD.front() <= i - k) minD.pop_front();
            while (!minD.empty() && values[minD.back()] >= values[i]) minD.pop_back();
            minD.push_back(i);
            
            while (!maxD.empty() && maxD.front() <= i - k) maxD.pop_front();
            while (!maxD.empty() && values[maxD.back()] <= values[i]) maxD.pop_back();
            maxD.push_back(i);
            
            // Median
            add(values[i]);
            if (i >= k) remove(values[i - k]);
            
            if (i >= k - 1) {
                int minVal = values[minD.front()];
                int maxVal = values[maxD.front()];
                int med = *small.rbegin();
                if (med == 0) result.push_back(0);
                else {
                    long long diff = (long long)maxVal - minVal;
                    long long instability = diff / med;
                    if (diff % med != 0 && ((diff ^ med) < 0)) {
                        instability--;
                    }
                    result.push_back(instability);
                }
            }
        }
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (cin >> n) {
        vector<int> remaining;
        int val;
        while (cin >> val) {
            remaining.push_back(val);
        }

        vector<int> values;
        int k = n / 2;  // Default

        // If we have exactly n values, use as values and k = n // 2
        if ((int)remaining.size() == n) {
            values.assign(remaining.begin(), remaining.end());
            k = n / 2;
        } else if ((int)remaining.size() == n + 1) {
            // First is k, rest are values
            k = remaining[0];
            values.assign(remaining.begin() + 1, remaining.begin() + n + 1);
        } else {
            // Try to parse k and values
            k = !remaining.empty() ? remaining[0] : n / 2;
            for (int i = 1; i <= n && i < (int)remaining.size(); i++) {
                values.push_back(remaining[i]);
            }
        }

        Solution solution;
        vector<long long> result = solution.windowInstability(values, k);
        for (int i = 0; i < (int)result.size(); i++) {
            if (i) cout << ' ';
            cout << result[i];
        }
        cout << "\n";
    }
    return 0;
}

File: dsa-problems/Queues/solutions/java/QUE-007-lab-window-instability.java
import java.util.*;

class Solution {
    public long[] windowInstability(int[] values, int k) {
        int n = values.length;
        long[] result = new long[n - k + 1];
        
        return solveWithDualPQ(values, k);
    }

    private long[] solveWithDualPQ(int[] values, int k) {
        int n = values.length;
        long[] result = new long[n - k + 1];
        Deque<Integer> minD = new ArrayDeque<>();
        Deque<Integer> maxD = new ArrayDeque<>();
        MedianFinder mf = new MedianFinder();

        for (int i = 0; i < n; i++) {
            // Min/Max
            while (!minD.isEmpty() && minD.peekFirst() <= i - k) minD.pollFirst();
            while (!minD.isEmpty() && values[minD.peekLast()] >= values[i]) minD.pollLast();
            minD.offerLast(i);

            while (!maxD.isEmpty() && maxD.peekFirst() <= i - k) maxD.pollFirst();
            while (!maxD.isEmpty() && values[maxD.peekLast()] <= values[i]) maxD.pollLast();
            maxD.offerLast(i);

            // Median
            mf.add(values[i]);
            if (i >= k) mf.remove(values[i - k]);

            if (i >= k - 1) {
                int minVal = values[minD.peekFirst()];
                int maxVal = values[maxD.peekFirst()];
                int median = mf.getMedian();
                if (median == 0) result[i - k + 1] = 0;
                else {
                    long diff = (long)maxVal - minVal;
                    long instability = diff / median;
                    if (diff % median != 0 && ((diff ^ median) < 0)) {
                        instability--;
                    }
                    result[i - k + 1] = instability;
                }
            }
        }
        return result;
    }

    static class MedianFinder {
        PriorityQueue<Integer> small = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> large = new PriorityQueue<>();
        // To handle lazy removal correctly, we track balance explicitly
        // Balance = small.size() - large.size() (considering only valid elements)
        // But since we can't know if a specific instance in heap is valid, we just track counts.
        // Let's use two TreeMaps (Multiset).
        
        // Re-implementation with TreeMap for correctness and simplicity
        TreeMap<Integer, Integer> smallMap = new TreeMap<>();
        TreeMap<Integer, Integer> largeMap = new TreeMap<>();
        int smallSize = 0;
        int largeSize = 0;

        void add(int val) {
            // Add to small first
            addMap(smallMap, val);
            smallSize++;
            
            // Move max of small to large
            int maxSmall = smallMap.lastKey();
            removeMap(smallMap, maxSmall);
            smallSize--;
            
            addMap(largeMap, maxSmall);
            largeSize++;
            
            // Rebalance: small should be >= large
            if (smallSize < largeSize) {
                int minLarge = largeMap.firstKey();
                removeMap(largeMap, minLarge);
                largeSize--;
                
                addMap(smallMap, minLarge);
                smallSize++;
            }
        }

        void remove(int val) {
            // Try to remove from small first
            if (smallMap.containsKey(val)) {
                // Check if it really belongs to small range
                // It must be <= small.lastKey()
                // Since small contains smaller elements, if val <= small.lastKey(), it's in small.
                // If val is in smallMap, we can just remove it?
                // Yes, because all elements in smallMap are <= all elements in largeMap.
                // So if val is in smallMap, it MUST be in the small partition.
                removeMap(smallMap, val);
                smallSize--;
            } else {
                removeMap(largeMap, val);
                largeSize--;
            }
            
            // Rebalance
            if (smallSize < largeSize) {
                int minLarge = largeMap.firstKey();
                removeMap(largeMap, minLarge);
                largeSize--;
                addMap(smallMap, minLarge);
                smallSize++;
            } else if (smallSize > largeSize + 1) {
                int maxSmall = smallMap.lastKey();
                removeMap(smallMap, maxSmall);
                smallSize--;
                addMap(largeMap, maxSmall);
                largeSize++;
            }
        }

        int getMedian() {
            return smallMap.lastKey();
        }

        void addMap(TreeMap<Integer, Integer> map, int val) {
            map.put(val, map.getOrDefault(val, 0) + 1);
        }

        void removeMap(TreeMap<Integer, Integer> map, int val) {
            int count = map.get(val);
            if (count == 1) map.remove(val);
            else map.put(val, count - 1);
        }
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (sc.hasNextInt()) {
            int n = sc.nextInt();
            int k = sc.nextInt();
            int[] values = new int[n];
            for (int i = 0; i < n; i++) {
                values[i] = sc.nextInt();
            }
            Solution solution = new Solution();
            long[] result = solution.windowInstability(values, k);
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < result.length; i++) {
                if (i > 0) sb.append(' ');
                sb.append(result[i]);
            }
            System.out.println(sb.toString());
        }
        sc.close();
    }
}
